sdk/python/feast/pyspark/historical_feature_retrieval_job.py

```diff
@@ -258,164 +621,158 @@ def join_entity_to_feature_tables(
-    """Retrieve batch features based on given configuration.
 
-    Args:
-        spark (SparkSession):
-            Spark session.
-        conf (Dict):
-            Configuration for the retrieval job, in json format. Sample configuration as follows:
+class SchemaError(Exception):
+    """
+    One or more columns in entity or feature table dataframe are either missing
+    or have the wrong data types
+    """
 
-    Returns:
-        DataFrame: Join result.
+    pass
 
-    Example:
-        sample_conf = {
-            "entity": {
-                "format": "csv",
-                "path": "file:///some_dir/customer_driver_pairs.csv",
-                "options": {"inferSchema": "true", "header": "true"},
-                "col_mapping": {
-                    "id": "driver_id"
-                },
-                "dtypes": {
-                    "driver_id": "integer"
-                }
-            },
-            "tables": [
-                {
-                    "format": "parquet",
-                    "path": "gs://some_bucket/bookings.parquet",
-                    "name": "bookings",
-                    "col_mapping": {
-                        "id": "driver_id"
-                    },
-                    "dtypes": {
-                        "driver_id": "integer"
-                    }
-                },
-                {
-                    "format": "avro",
-                    "path": ""s3://some_bucket/transactions.parquet"",
-                    "name": "transactions",
-                },
-            ],
-            "queries": [
-                {
-                    "table": "transactions",
-                    "features": ["daily_transactions"],
-                    "join": ["customer_id"],
-                    "max_age": 172800,
-                },
-                {
-                    "table": "bookings",
-                    "features": ["completed_bookings"],
-                    "join": ["driver_id"],
-                },
-            ],
-            "output":
-                "format": "parquet"
-                "path": "gs://some_bucket/output.parquet"
-        }
 
-        The values for the `format` and `path` should be recognizable by the Spark cluster where the job
-        is going to run on. For example, if you specify `bigquery` as input format, then you should ensure
-        that the Spark Big Query connector is installed on the cluster. Like wise, s3a connector is required
-        for Amazon S3 path.
+def retrieve_historical_features(
+    spark: SparkSession,
+    entity_source: Source,
+    feature_tables_sources: List[Source],
+    feature_tables: List[FeatureTable],
+) -> DataFrame:
+    """Retrieve historical features based on given configurations.
 
-        `options` is optional. If present, the options will be used when reading / writing the input / output.
+    Args:
+        spark (SparkSession): Spark session.
+        entity_source (Source): Entity data source.
```

Comment:
I don't think "entity data source" is a concept that has been used before in the codebase, maybe worth adding an explanation what it is. Or maybe call it `entity_df_source`? It is somewhat confusing that "entities" is used interchangeably for the entity objects themselves and the dataframe for point-in-time joins.